src/conduits/RwaOutputConduit3.sol

```diff
@@ -0,0 +1,380 @@
+        to = who;
+        emit Pick(who);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Method to swap DAI contract balance to GEM through PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     */
+    function push() external isMate {
+        _doPush(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of DAI into GEM through the PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     * @param wad DAI amount.
+     */
+    function push(uint256 wad) external isMate {
+        _doPush(wad);
+    }
+
+    /**
+     * @notice Flushes out any DAI balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out the specified amount of DAI to the `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param wad DAI amount.
+     */
+    function quit(uint256 wad) external isMate {
+        _doQuit(wad);
+    }
+
+    /**
+     * @notice Flushes out any GEM balance to `usr` address.
+     * @param usr Destination address
+     * @dev Can be called only by admin.
+     */
+    function quitGem(address usr) external auth {
+        gem.transfer(usr, gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Calculate amount of GEM received for swapping DAI through PSM.
+     * @param wad DAI amount.
+     */
+    function daiToGem(uint256 wad) public view returns (uint256) {
+        return mul(wad, WAD) / add(WAD, psm.tout()) / to18ConvertionFactor;
+    }
+
+    /**
+     * @notice Swaps the specified amount of DAI into GEM through the PSM and push it to the recipient address.
+     * @param wad DAI amount.
+     */
+    function _doPush(uint256 wad) internal {
+        require(to != address(0), "RwaOutputConduit3/to-not-picked");
+
+        // We might lose some dust here because of rounding errors. I.e.: USDC has 6 dec and DAI has 18.
+        uint256 gemAmount = daiToGem(wad);
+        require(gemAmount > 0, "RwaOutputConduit3/insufficient-swap-gem-amount");
+
+        uint256 prevGemBalance = gem.balanceOf(address(this));
+
+        psm.buyGem(address(this), gemAmount);
+
+        uint256 gemBalance = gem.balanceOf(address(this));
+        uint256 gemPushAmt = sub(gemBalance, prevGemBalance);
+        address _to = to;
+
+        to = address(0);
+        gem.transfer(_to, gemPushAmt);
+
+        emit Push(_to, gemPushAmt);
+    }
+
+    /**
+     * @notice Flushes out the specified amount of DAI to `quitTo` address.
+     * @param wad The DAI amount.
+     */
+    function _doQuit(uint256 wad) internal {
```

Comment:
as discussed I think the pattern of having internal functions right after their wrappers is nicer - https://github.com/makerdao/dss-vest/blob/master/src/DssVest.sol#L232

Reply:
At CES we follow a different convention for order of declaration in contracts:

1. `using` statements
2. Struct declarations
3. `public constant` variables
4. `public immutable` variables
5. `public immutable` variables
6. `internal/private constant` variables*
7. `internal/private immutable` variables*
8. `public` primitive type variables*
9. `public` complex types variables (mappings, arrays, etc.)*
10. `private` primitive type variables*
11. `private` complex types variables (mappings, arrays, etc.)*
12. Events
13. Modifiers
14. Constructor
15. `public/external (payable)`  methods
16. `public/external view`  methods
17. `public/external pure`  methods
18. `private/internal (payable)`  methods
16. `private/internal view`  methods
17. `private/internal pure`  methods
18. Code unrelated to the main contract subject (i.e.: Math functions).

<sup>*</sup> There are exceptions to these rules, specially when they impact the storage layout of the contracts that are common to MakerDAO contract (i.e.: `wards` in slot `0`).

We prioritize 2 things with this structure:
1. It's easy to figure out in which slot each storage variable is located.
2. Having the externally accessible functions shown first gives the reader a good sense of what the contract does without having to bother too much with implementation details.

While we do appreciate the "group by feature" mentality, since we try take a more modular approach and try to keep our smart contracts short and concise, the gains from using it are mostly foregone.

One could argue that this approach leads to a lot of jumping around when reviewing the code. However in this specific case the wrappers are literal one-liners, so there is no need to go back and forth when reading it.

Reply:
Yes I guess it depends on what we consider more important, internal guidelines or trying to keep more consistency within all the MakerDAO contracts.

Reply:
TBH MakerDAO contracts are not that consistent, we've seem lots of deviations from what would be considered the standard and there's no official documentation regarding that.

This structure is something that we intensively discussed internally and we are open to make some concessions when there is a good reason to do so.

However, when it comes to a simple matter of preference or because a specific contract is written differently, we will stick with the standards that we defined internally.

Reply:
It's true the contracts form have evolved, for example from the old pattern of putting all the wards related together to now having by group of code (storages separated from modifier and functions). And it is true we do not have strong guidelines either but there are a couple of things that come natural from prev contracts (like math above).
Other stuff like the regular private functions, it has been more subject to the writer, we also tried to group together (above though), but in general makes it more annoying for reviewing.
Anyway, stick with whatever you want, we should probably treat this as another third party code.

Reply:
There should probably be a separate discussion on whether the new conduits should be regarded as external integrations contracts.
Closing this as we voiced our opinion.